src/script_opt/Stmt.cc

```diff
@@ -971,7 +951,7 @@ void WhenInfo::UpdateIDs(Reducer* c)
 
 StmtPtr WhenStmt::Duplicate()
 	{
-	return SetSucc(new WhenStmt(new WhenInfo(wi)));
+	return SetSucc(new WhenStmt(std::make_shared<WhenInfo>(wi.get())));
```

Comment:
Making a second shared_ptr here but pointing at the same block of memory can be dangerous one of the two shared_ptrs goes fully out of scope and deletes it's memory. Is there any reason to not just pass the shared pointer down? It'd get `move`d but `WhenStmt`'s constructor already took a copy of the shared pointer (and updated the ref count) because it was passed by value.

Reply:
FYI @timwoj this is actually different than it looked - the point isn't to duplicate the pointer, it's to create a new `WhenInfo` object that's initialized using the existing one.